### ISSUE TITLE
curl command with both "--head" and "--request GET" options

### ISSUE THREAD

**Issue by thuctnguyen** [Homebrew/brew#10817]
### `brew config` output

brew config
HOMEBREW_VERSION: 3.0.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: ee52b19179d056f3dd2fdf9c52214d3d51dbb9d9
Last commit: 7 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: a4e6f38e850654f729fea87ea147362fb0f1c295
Core tap last commit: 5 days ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: octa-core 64-bit kabylake
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.2.2-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A

### `brew doctor` output

brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed with `brew diy`.
You should find replacements for the following formulae:
...........

### 

- [X] The `brew doctor` above contains no "Warning" lines.

### What were you trying to do (and why)?

brew fetch --debug xyz

I was debugging "connection reset" errors on the server side. I turned on debugging to see the actual curl commands.

### What happened (include all command output)?

Here's one of the curl commands in the output:

/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/3.0.5\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.2.2\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --silent --head --request GET url-to-tarball

You can see in that command that both "--head" and "--request GET" options were provided. For "--head" curl read the headers off of the response and closed the connection. However, since "--request GET" was also provided, the server continued to send back the response. curl then responded with RST packets which caused the server to output "connection reset".

### What did you expect to happen?

"--request GET" shouldn't be provided in this case.

### Step-by-step reproduction instructions (by running `brew` commands)

brew fetch --debug xyz

**Comment by carlocab** [comment]
Can you give an example of a formula where passing both flags causes problems? It doesn't seem to for most formulae. For example:
```
❯ brew fetch --debug hello
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/hello.rb
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/3.0.4-75-g300f6f1-dirty\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.2.2\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --silent --head --request GET https://homebrew.bintray.com/bottles/hello-2.10.big_sur.bottle.tar.gz
==> Downloading https://homebrew.bintray.com/bottles/hello-2.10.big_sur.bottle.tar.gz
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/3.0.4-75-g300f6f1-dirty\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.2.2\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --range 0-1 --dump-header - --write-out \%\{http_code\} --output /dev/null https://homebrew.bintray.com/bottles/hello-2.10.big_sur.bottle.tar.gz
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/3.0.4-75-g300f6f1-dirty\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.2.2\)\ curl/7.64.1 --header Accept-Language:\ en --fail --progress-bar --retry 3 --location --remote-time --continue-at 0 --output /Users/chisquared/Library/Caches/Homebrew/downloads/dd816ddcb29e99b3c2d0f3db4be2e8c5b2dfcb0006c0ffc12eb0cf30f9e626d9--hello-2.10.big_sur.bottle.tar.gz.incomplete https://homebrew.bintray.com/bottles/hello-2.10.big_sur.bottle.tar.gz
######################################################################## 100.0%
Downloaded to: /Users/chisquared/Library/Caches/Homebrew/downloads/dd816ddcb29e99b3c2d0f3db4be2e8c5b2dfcb0006c0ffc12eb0cf30f9e626d9--hello-2.10.big_sur.bottle.tar.gz
SHA256: 69489ae397e4645127aa7773211310f81ebb6c99e1f8e3e22c5cdb55333f5408
```

**Comment by thuctnguyen** [comment]
In your example, you can see this problem in this command

> /usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/3.0.4-75-g300f6f1-dirty\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.2.2\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --silent --head --request GET https://homebrew.bintray.com/bottles/hello-2.10.big_sur.bottle.tar.gz

For us it's a formula that we build in house. It downloads the tarball from an on-prem instance of Artifactory. We see lots of "connection reset" errors on Artifactory side. On the client/Homebrew side, things work just find as curl would just ignore the additional response that Artifactory sends over.

**Comment by carlocab** [comment]
>you can see this problem in this command

I can see that both flags are passed. What I can't see is the problem, because the bottle was fetched properly for me. Am I understanding you correctly that my `brew fetch` operation actually generated a bunch of errors on Bintray's side of things?

**Comment by Bo98** [comment]
Just checked on this. It's intentional because of this issue: https://github.com/Homebrew/brew/issues/5895

**Comment by jonchang** [comment]
You may want to write a custom download strategy that better fits your needs, since this is for a private repository and not something in homebrew/core: https://rubydoc.brew.sh/AbstractFileDownloadStrategy.html

**Comment by Bo98** [comment]
> https://rubydoc.brew.sh/AbstractFileDownloadStrategy.html

Side note: I'm surprised that's private API. We should probably consider making it public if it's something we recommend taps use.

**Comment by jonchang** [comment]
Basically everything in Homebrew outside of the Formula class is a private API. We deprecate stuff pretty aggressively, for better or for worse.

**Comment by Bo98** [comment]
Sure, but we skip deprecation entirely for stuff truly private.

**Comment by thuctnguyen** [comment]
The fix for #5895 is interesting. I think `curl` should have been fixed to support "--head" properly rather than adding "--request GET" to the command.

I emailed `curl-users` to inquire about this use case where both options were provided. I was sent to this old [blog post](https://daniel.haxx.se/blog/2015/09/11/unnecessary-use-of-curl-x/) discussing various unnecessary usages of `curl` options.

While everything seems OK on the client/Homebrew side, it amounts to downloading the same file twice from the server side. A HEAD request should just return the headers and nothing else. Otherwise, you should just do a GET to get both the headers and the body.

**Comment by Bo98** [comment]
Yeah, I understand the concern on the server side.

The issue in #5895 is likely a serverside issue itself. In fact, the specific example in question there has now been fixed serverside. There way well be something else in `homebrew-core` though that is currently relying on this - it can be a tricky game.

We can only do what we can with the tools provided. History has shown us there is various badly configured web servers out there, and ultimately we have to workaround them as there's just some that will never be fixed.
